### PR TITLE
Remove `_get_ν`

### PR DESCRIPTION
**Pull request by devmotion**
Fix the bug introduced in #425 (see https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/450).

Removes the workaround.

### CODE REVIEW

**Comment by theogf** [comment]
Is it possible to test this?

**Comment by devmotion** [comment]
> So the workaround is tested by virtue of the Zygote AD test on MaternKernel now also passing.

The main problem is that tests passed successfully for https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/452/commits/08437ccc44fa134b86eb447f59d0bdfe651d6acd even though the pullback is clearly incorrect. So I was wondering if the pullback was evaluated at all.

File: src/basekernels/matern.jl [hunk 1]
```diff
@@ -37,8 +37,18 @@ MaternKernel(; nu::Real=1.5, ν::Real=nu, metric=Euclidean()) = MaternKernel(ν,
 
 @functor MaternKernel
 
+# workaround for Zygote
+# unclear why it's needed but it is fine since it's stated officially that we don't support differentiation with respect to ν
```
Comment (st--) [inline src/basekernels/matern.jl:41]: Without this workaround, the Zygote AD test fails. You can see this in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/425/commits/30385acdee781a48d5960fe190b87c2e6c43aad0 and the corresponding build https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/runs/6002685848?check_suite_focus=true

**Comment by st--** [comment]
> The main problem is that tests passed successfully for [08437cc](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/commit/08437ccc44fa134b86eb447f59d0bdfe651d6acd) even though the pullback is clearly incorrect.

What was incorrect about the pullback?

> So I was wondering if the pullback was evaluated at all.

It was, I only added it because without Zygote errored, see https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/452#discussion_r849550385

**Comment by devmotion** [comment]
> What was incorrect about the pullback?

It returned the wrong number of outputs (was missing `NoTangent()` derivative for the function itself).

> Without this workaround, the Zygote AD test fails. You can see this in [30385ac](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/commit/30385acdee781a48d5960fe190b87c2e6c43aad0) and [the corresponding build](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/runs/6002685848?check_suite_focus=true)

The problem is just that https://github.com/FluxML/Zygote.jl/blob/95e10218f0187a839ac0539fefcbf39c727bf068/src/lib/array.jl#L30 does not support `NotImplemented`. This should have been fixed in Zygote, introducing workarounds in KernelFunctions is the wrong approach.

**Comment by st--** [comment]
> This should have been fixed in Zygote, introducing workarounds in KernelFunctions is the wrong approach.

Yeah, I completely agree, but unfortunately it _hasn't_ been fixed in Zygote, and I think "workaround + AD is working" is still preferable over "no workaround + AD is broken".

**Comment by devmotion** [comment]
Instead of making a PR to KernelFunctions introducing a workaround you could have made one for Zygote that fixes the issue :thinking:

**Comment by st--** [comment]
> Instead of making a PR to KernelFunctions introducing a workaround you could have made one for Zygote that fixes the issue :thinking:

I managed to figure out how to fix the issue with a workaround in KernelFunctions. That doesn't mean I would immeditely know how to identify & fix the underlying issue upstream. If I _did_ know how to fix it there, I would have, as I have in various other cases. You seem to have an amazing overview as well as deep understanding of a lot of different packages. That's unfortunately not true of everyone, and I'm assuming you're not intending to but you give the impression of being rather disparaging of other people who're not at your level.

**Comment by devmotion** [comment]
IMO usually it is very helpful to open at least Github issues in cases where it is not immediately clear how to fix the issue. Workarounds can be acceptable if they are fixed upstream eventually, but I assume it is unlikely that they are fixed if nobody makes the Zygote developers aware of it.

Generally, I would have given you this advice in your PR but as desired I don't review your PRs anymore.

**Comment by st--** [comment]
@devmotion I've pared it down to a minimal reproducible example: https://github.com/FluxML/Zygote.jl/issues/1204 but I can't comment on anything else, if you've got further thoughts on what the issue is / what's needed to fix it, please comment there.

**Comment by devmotion** [comment]
Test errors seem unrelated? I didn't change anything related to them, I think.

**Comment by willtebbutt** [comment]
> Test errors seem unrelated? I didn't change anything related to them, I think.

Agreed. Maybe we need to relax a tolerance somewhere?

**Comment by st--** [comment]
> > Test errors seem unrelated? I didn't change anything related to them, I think.
> 
> Agreed. Maybe we need to relax a tolerance somewhere?

Yes, I think it's due to the changes in #447; the tests are stochastic (if you just re-run the failed tests, they might pass the next time - I've seen that in #454 as well which has zero code changes). In #447 I reduced the size of the kernel matrix evaluations; the isapprox() call (set with atol=1e-9, rtol=1e-9) is on the arrays as a whole, which does not look at individual elements, but instead checks whether the norm of the difference is within the tolerances, relative to the max of the norm of the individual matrices (see [discussion](https://discourse.julialang.org/t/unit-test-equality-for-arrays/8138/15)). For the random inputs we're using in the tests, the norm of the kernel matrix is linear in the dimensions. So by reducing the size from ~1000 to ~10 we're reducing the norm by a factor ~100, so yes that suggests to me we should increase the rtol correspondingly.

**Comment by devmotion** [comment]
Shouldn't the scale be irrelevant if we use `rtol`? I think it's problematic that we don't use `rtol` in most of the tests but only in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/8e805ef25d745ab391e4ad424e690507e02cd042/src/test_utils.jl#L75-L76. Probably because otherwise tests failed - which IMO, however, would not indicate a problem with `rtol` per se but that the default already was already too low even before the recent changes. I completely agree with https://discourse.julialang.org/t/unit-test-equality-for-arrays/8138/12?u=devmotion that in general you want `rtol` since it is completely unclear what a meaningful `atol` is in general. Specifying a small `atol` as well mainly helps with comparisons that involve exact 0s which would fail otherwise.

**Comment by st--** [comment]
> I think it's problematic that we don't use `rtol` in most of the tests

Yes, I agree. Probably an oversight in the original test code of not realizing that by only setting atol you end up setting rtol=0...

**Comment by devmotion** [comment]
Apparently https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/455 was not sufficient.

**Comment by devmotion** [comment]
Bump.

**Comment by devmotion** [comment]
Bump.

**Comment by theogf** [comment]
Could we keep the global atol and rtol variables?

**Comment by devmotion** [comment]
I don't mind. It just seemed unnecessary to have global variables that are used only in a single place, and clearer (ie easier to spot) if the values are just used as default keyword arguments in the function definition.

**Comment by devmotion** [comment]
@theogf Did you see my last comment? Would you like it to be reverted?

**Comment by theogf** [comment]
I think keeping the global const is a bit more elegant and easily findable since it's in the same file. But again I do not mind, I'll just accept the PR and do what you want 😅

**Review by theogf — APPROVED** [review]

**Pull request merged**